### PR TITLE
pcapgo: add OpenEthernet and a full capture example

### PR DESCRIPTION
**Pull request by stapelberg**
I just implemented this in a project of mine and figured it might make a good addition to the pcapgo package :)

### CODE REVIEW

**Comment by stapelberg** [comment]
Friendly ping? :)

File: pcapgo/capture.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
+package pcapgo
```
Comment (gconnell) [inline pcapgo/capture.go]: please copy over copyright notices to top of new files.

File: pcapgo/capture.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+// ReadPacketData implements gopacket.PacketDataSource.
+func (h *EthernetHandle) ReadPacketData() ([]byte, gopacket.CaptureInfo, error) {
+	b := make([]byte, 4096) // TODO(correctness): how much space do we need?
+	n, _, err := h.ReadFrom(b)
+	if err != nil {
+		return nil, gopacket.CaptureInfo{}, err
+	}
+	packet := gopacket.NewPacket(b[:n], layers.LayerTypeEthernet, gopacket.Default)
```
Comment (gconnell) [inline pcapgo/capture.go]: Don't create a new packet and return it.  Just return b[:n], and gopacket's https://github.com/google/gopacket/blob/master/packet.go#L794 will wrap it into a packet.
With the current code, a full packet decode happens, then the entire packet object is thrown out.

File: pcapgo/capture.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	data := packet.Data()
+	return data, gopacket.CaptureInfo{
+		CaptureLength: len(data),
+		Length:        len(data),
+		Timestamp:     time.Now(),
+	}, nil
+}
+
+// OpenEthernet implements pcap.OpenLive for ethernet interfaces only.
+func OpenEthernet(ifname string) (*EthernetHandle, error) {
```
Comment (gconnell) [inline pcapgo/capture.go]: to make things perfectly clear, maybe "NewEthernetHandle(...)"

File: pcapgo/capture_test.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
+package pcapgo_test
```
Comment (gconnell) [inline pcapgo/capture_test.go]: copyright notice here plz

File: pcapgo/capture_test.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+import (
+	"log"
+	"os"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"github.com/google/gopacket/pcapgo"
+)
+
+func Example_captureEthernet() {
```
Comment (gconnell) [inline pcapgo/capture_test.go]: nit: CaptureEthernet

File: pcapgo/capture_test.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+	f, err := os.Create("/tmp/lo.pcap")
```
Comment (gconnell) [inline pcapgo/capture_test.go]: I've never seen this before... why does it work?  Why does this file translate to net.InterfaceByName working?  Inquiring minds like mine want to know :)

File: pcapgo/capture.go (hunk 1 above)
Comment (stapelberg) [reply pcapgo/capture.go]: Done

File: pcapgo/capture.go (hunk 2 above)
Comment (stapelberg) [reply pcapgo/capture.go]: Whoops, of course. Done.

File: pcapgo/capture_test.go (hunk 4 above)
Comment (stapelberg) [reply pcapgo/capture_test.go]: Done.

File: pcapgo/capture.go (hunk 3 above)
Comment (stapelberg) [reply pcapgo/capture.go]: Done.

**Comment by stapelberg** [comment]
I addressed some of your comments, but got to run now :) I’ll try to address the remaining ones later.

Thanks for your review!

File: pcapgo/capture_test.go (hunk 5 above)
Comment (stapelberg) [reply pcapgo/capture_test.go]: This is a load-bearing lowercase “c”: if I make it Example_CaptureEthernet, godoc won’t show the example.

File: pcapgo/capture_test.go (hunk 6 above)
Comment (stapelberg) [reply pcapgo/capture_test.go]: Not sure I follow. The filename has nothing to do with the rest of the code? NewEthernetHandle takes an ifname, which is passed to net.InterfaceByName, and we specify ifname="lo".

**Comment by stapelberg** [comment]
All addressed now. PTAL.

File: pcapgo/capture_test.go (hunk 5 above)
Comment (gconnell) [reply pcapgo/capture_test.go]: lolwut

**Comment by gconnell** [comment]
Looks like Travis is having some trouble: https://travis-ci.org/google/gopacket/builds/425491900?utm_source=github_status&utm_medium=notification

I believe due to the import of mdlayher/raw in pcapgo.  

I think this would be fixed by adding:
  - go get github.com/google/gopacket/pcapgo
to the set of 'go get' calls in the Travis config?

**Comment by stapelberg** [comment]
Woops, didn’t realize the package names were enumerated in the travis configuration. Added now.

**Comment by stapelberg** [comment]
Thanks!

**Pull request merged**